website/docs/language/modules/develop/refactoring.html.md

```diff
@@ -0,0 +1,475 @@
+---
+layout: "language"
+page_title: "Refactoring"
+sidebar_current: "docs-modules-recactoring"
+description: Making backward-compatible changes to modules already in use.
+---
+
+# Refactoring
+
+~> **Note:** Explicit refactoring declarations with `moved` blocks will work
+only in Terraform v1.1 and later. When working with earlier Terraform versions,
+or for refactoring actions too complex to express as `moved` blocks, you could
+use
+[the `terraform state mv` CLI command](/docs/cli/commands/state/mv.html)
+as a separate step instead.
+
+In shared modules and long-lived configurations you may eventually outgrow
+your initial module heirarchy and resource naming and need to adopt a new
+strategy. For example, you might decide that what was previously one
+child module makes more sense as two separate modules, with each of the
+resources moving to one of them.
+
+Terraform compares previous state with new configuration, correlating by
+each module or resource's unique address. Therefore _by default_ Terraform will
+understand moving or renaming an object as an intent to destroy the object
+at the old address and to create a new object at the new address.
+
+You can use `moved` blocks in your configuration to record where you've
+historically moved or renamed an object, in which case Terraform will instead
+treat an existing object at the old address as if it now belongs to the new
+address.
```

Comment:
Good point!

I think the "why is this good" can vary between situations, so I'll have a think about how to write this in a more concise way, but in the interests of capturing the answer somewhere we can refer to it for later edits, reasons might include:

* As you suggested, it could be time-consuming to destroy and recreate some kinds of objects.
* The brief period where there are fewer (or no) objects of a particular type, or in some cases _too many_ objects of a particular type (if we're using `create_before_destroy`) can potentially cause total downtime or problematic backpressure on other subsystems. For example, adding new members to a Consul cluster will cause the cluster to temporarily thrash as it replicates data from the existing servers to the new servers, which might hurt performance for systems depending on that cluster.
* Some objects are not trivially replaceable. For example, if we have a resource representing a database server and we replace the database server then we'll probably be replacing a server which has data in it with a server that has _no_ data in it.

Perhaps the general way to talk about this is the idea that there are lots of infrastructure object types where replacing them requires careful planning and possibly coordination with other teams, and so module users and authors would typically prefer that _just_ upgrading to a new version of a module (without making other changes) would not involve replacing any existing objects.

I'll continue to ponder this, though if what I've said here gives _you_ any ideas about how to word this then of course I'd love to hear them!

Reply:
I will think about this, too! Thank you as always for the awesome explanation :-)